### PR TITLE
Set read timeout for fetching IMDSv2 token

### LINKED ISSUE #104244: Fetching IMDSv2 token should have timeout for read
We configure connectTimeout for fetching token from IMDSv2's token API.

https://github.com/elastic/elasticsearch/blob/1e1d15129a51a5a4ec24c4a862b5313ff149d38b/plugins/discovery-ec2/src/main/java/org/elasticsearch/discovery/ec2/AwsEc2Utils.java#L42

However, we do not have timeout for reading from the connection. Therefore, the read can get stuck* indefinitely

https://github.com/elastic/elasticsearch/blob/1e1d15129a51a5a4ec24c4a862b5313ff149d38b/plugins/discovery-ec2/src/main/java/org/elasticsearch/discovery/ec2/AwsEc2Utils.java#L49

Since we rely on error or empty respnose from the token API to fallback to IMDSv1 (#84410), stucking at reading prevents the fallback mechanism to work.

[*] One reason for read to stuck is that the request needs more than 1 hop https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instance-metadata-v2-how-it-works.html

### PR DESCRIPTION
**Pull request by arteam**
Resubmit of #104397 ~with the fixed signature of setting the environment variable in the build~ without setting `AWS_METADATA_SERVICE_TIMEOUT` randomly in the build.

> Use the timeout set by AWS_METADATA_SERVICE_TIMEOUT environment variable both as connect and read timeout analogous to the AWS SDK.
See https://docs.aws.amazon.com/sdkref/latest/guide/feature-ec2-instance-metadata.html

Resolves #104244

### CODE REVIEW

File: plugins/discovery-ec2/build.gradle [hunk 1]
```diff
@@ -102,6 +102,9 @@ tasks.named("test").configure {
   } else {
     nonInputProperties.systemProperty 'java.security.policy', "file://${buildDir}/tmp/java.policy"
   }
+  if (BuildParams.random.nextBoolean()) {
+    environment 'AWS_METADATA_SERVICE_TIMEOUT', '1'
```
[the remark this answers is not in the document]
Comment (kingherc) [reply plugins/discovery-ec2/build.gradle:106]: Is there a reason for the random boolean? I thought you wanted to make the test quick in CI and that's why we set the default to 1 here.

**Review by mark-vieira — CHANGES_REQUESTED** [review]

**Comment by arteam** [comment]
@mark-vieira That makes sense, I've removed the randomness from the build, so we will just be testing the client the default timeout of 2s.

**Review by mark-vieira — APPROVED** [review]

File: plugins/discovery-ec2/build.gradle (hunk 1 above)
[the remark this answers is not in the document]
Comment (ywangd) [reply plugins/discovery-ec2/build.gradle]: The code computes the timeout value based on the environment variable or falls back to a hardcoded default if the environment variable does not exist. Ideally we should be testing this behaviour as well. I see introducing this with Gradle build file is not prefered. The other alternative is dependency injection like how it is done for `CustomWebIdentityTokenCredentialsProvider`
https://github.com/elastic/elasticsearch/blob/c8fc737846e927f3caa134df395f0f9d71af66f4/modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Service.java#L332-L335

But that requires some cascading refactoring and may _not_ be worth the effort. Hence I am ok with just testing the default behaviour in this PR.

**Review by ywangd — APPROVED** [review]
LGTM

**Review by kingherc — APPROVED** [review]
LGTM

**Pull request merged**